x-pack/plugins/licensing/server/plugin.ts

```diff
@@ -4,145 +4,170 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import { Observable } from 'rxjs';
-import { first, map } from 'rxjs/operators';
+import { Observable, Subject, Subscription, merge, timer } from 'rxjs';
+import { take, takeUntil } from 'rxjs/operators';
 import moment from 'moment';
+import { createHash } from 'crypto';
+
 import {
   CoreSetup,
   CoreStart,
   Logger,
-  Plugin as CorePlugin,
+  Plugin,
   PluginInitializerContext,
+  IClusterClient,
 } from 'src/core/server';
-import { Poller } from '../../../../src/core/utils/poller';
-import { LicensingConfigType, LicensingPluginSetup, ILicense } from './types';
-import { LicensingConfig } from './licensing_config';
-import { License } from './license';
+
+import { ILicense, LicensingPluginSetup, PublicLicense } from '../common/types';
+import { License } from '../common/license';
+import { createLicenseUpdate } from '../common/license_update';
+
+import { ElasticsearchError, RawLicense, RawFeatures } from './types';
+import { LicenseConfigType } from './licensing_config';
 import { createRouteHandlerContext } from './licensing_route_handler_context';
 
-declare module 'src/core/server' {
-  interface RequestHandlerContext {
-    licensing: {
-      license: ILicense;
+function normalizeServerLicense(license: RawLicense): PublicLicense['license'] {
+  return {
+    uid: license.uid,
+    type: license.type,
+    expiryDateInMillis: license.expiry_date_in_millis,
+    status: license.status,
+  };
+}
+
+function normalizeFeatures(rawFeatures: RawFeatures) {
+  const features: PublicLicense['features'] = {};
+  for (const [name, feature] of Object.entries(rawFeatures)) {
+    features[name] = {
+      isAvailable: feature.available,
+      isEnabled: feature.enabled,
     };
   }
+  return features;
 }
 
-export class Plugin implements CorePlugin<LicensingPluginSetup> {
+function sign(
+  licenseJSON: Partial<Pick<PublicLicense, 'license' | 'features'>>,
+  error: string = ''
+) {
+  return createHash('md5')
+    .update(JSON.stringify({ ...licenseJSON, error }))
+    .digest('hex');
```

Comment:
Question because not sure about JS implementation: Are we sure `{ ...licenseJSON, error }` will be consistent in key ordering here?

Reply:
Yes, object key order is consistent starting with ES2015.

Reply:
Wow, I'm so outdated

Reply:
While I think key order in objects is stable, it doesn't appear that key order in `JSON.stringify` implemenations is necessarily stable. In other parts of Kibana we use the `json-stable-stringify` package to ensure this is stable. I think we should use this here too.